python/cudf_polars/cudf_polars/utils/config.py

```diff
@@ -0,0 +1,137 @@
+# SPDX-FileCopyrightText: Copyright (c) 2024-2025, NVIDIA CORPORATION & AFFILIATES.
+# SPDX-License-Identifier: Apache-2.0
+
+"""Config utilities."""
+
+from __future__ import annotations
+
+import copy
+import json
+from typing import TYPE_CHECKING, Any
+
+if TYPE_CHECKING:
+    from typing_extensions import Self
+
+__all__ = ["ConfigOptions"]
+
+
+class ConfigOptions:
+    """
+    GPUEngine configuration-option manager.
+
+    This is a convenience class to help manage the nested
+    dictionary of user-accessible `GPUEngine` options.
+    """
+
+    __slots__ = ("_hash_value", "config_options")
+    _hash_value: int
+    config_options: dict[str, Any]
+    """The underlying (nested) config-option dictionary."""
+
+    def __init__(self, options: dict[str, Any]):
+        self.validate(options)
+        self.config_options = options
+
+    def set(self, name: str, value: Any) -> Self:
+        """
+        Set a user config option.
+
+        Nested dictionary keys should be separated by periods.
+        For example::
+
+            >>> options = options.set("parquet_options.chunked", False)
+
+        Parameters
+        ----------
+        name
+            Period-separated config name.
+        value
+            New config value.
+        """
+        options = config_options = copy.deepcopy(self.config_options)
+        keys = name.split(".")
+        for k in keys[:-1]:
+            assert isinstance(options, dict)
+            if k not in options:
+                options[k] = {}
+            options = options[k]
+        options[keys[-1]] = value
+        return type(self)(config_options)
+
+    def get(self, name: str, *, default: Any = None) -> Any:
+        """
+        Get a user config option.
+
+        Nested dictionary keys should be separated by periods.
+        For example::
+
+            >>> chunked = config_options.get("parquet_options.chunked")
+
+        Parameters
+        ----------
+        name
+            Period-separated config name.
+        default
+            Default return value.
+
+        Returns
+        -------
+        The user-specified config value, or `default`
+        if the config is not found.
+        """
+        options = self.config_options
+        keys = name.split(".")
+        for k in keys[:-1]:
+            assert isinstance(options, dict)
+            options = options.get(k, {})
+        return options.get(keys[-1], default)
+
+    def __hash__(self) -> int:
+        """Hash a ConfigOptions object."""
+        try:
+            return self._hash_value
+        except AttributeError:
+            self._hash_value = hash(json.dumps(self.config_options))
+            return self._hash_value
```

Comment:
This caching is unsafe because we can `set` values which would invalidate the hash. Should we remove `set`?

Reply:
It is *possible* to invalidate the hash if you extract a dictionary from the config and manually set it to something different.  However, `set` will always return a new `ConfigOptions` object (without the `_hash_value` attribute). Given that this is meant as an internal (and probably temporary) convenience API, this behavior seems "okay" to me.

I am currently relying on the `set` functionality to disable chunked parquet reading, so hopefully we can keep that?

Reply:
To put it another way, it is "safe" to use `set`, and "unsafe" to change a config value manually (without `set`). Therefore, I'd prefer to keep the safe API around since I do need to change the config in at least one place.

Reply:
Aha, sorry, I missed that. In which case, this is fine.